src/WebJobs.Script.WebHost/Security/KeyManagement/SimpleKubernetesClient.cs

```diff
@@ -0,0 +1,307 @@
+        {
+            if (string.IsNullOrEmpty(KubernetesObjectName) && FileUtility.DirectoryExists(KubernetesSecretsDir))
+            {
+                return await GetFromFiles(KubernetesSecretsDir);
+            }
+            else if (!string.IsNullOrEmpty(KubernetesObjectName))
+            {
+                return await GetFromApiServer(KubernetesObjectName);
+            }
+            else
+            {
+                throw new InvalidOperationException($"{nameof(KubernetesSecretsRepository)} requires setting {EnvironmentSettingNames.AzureWebJobsKubernetesSecretName} or mounting secrets to {KubernetesSecretsDir}");
+            }
+        }
+
+        public async Task UpdateSecrets(IDictionary<string, string> data)
+        {
+            (var url, var isSecret) = await GetObjectUrl(KubernetesObjectName);
+            await CreateIfDoesntExist(url, isSecret);
+
+            data = isSecret
+                ? data.ToDictionary(k => k.Key, v => Convert.ToBase64String(Encoding.UTF8.GetBytes(v.Value)))
+                : data;
+
+            using (var request = await GetRequest(HttpMethod.Patch, url, new[] { new { op = "replace", path = "/data", value = data } }, "application/json-patch+json"))
+            {
+                var response = await _httpClient.SendAsync(request);
+                response.EnsureSuccessStatusCode();
+            }
+        }
+
+        public void OnSecretChange(Action callback)
+        {
+            _watchCallback = callback;
+        }
+
+        private async Task RunWatcher()
+        {
+            if (string.IsNullOrEmpty(KubernetesObjectName) && FileUtility.DirectoryExists(KubernetesSecretsDir))
+            {
+                _fileWatcher = new AutoRecoveringFileSystemWatcher(KubernetesSecretsDir);
+                _fileWatcher.Changed += (object sender, FileSystemEventArgs e)
+                    => _watchCallback?.Invoke();
+            }
+            else if (!string.IsNullOrEmpty(KubernetesObjectName))
+            {
+                (var url, _) = await GetObjectUrl(KubernetesObjectName, watchUrl: true);
+                using (var noTimeoutClient = CreateHttpClient())
+                using (var request = await GetRequest(HttpMethod.Get, url))
+                {
+                    while (!_disposed)
+                    {
+                        noTimeoutClient.Timeout = TimeSpan.FromMilliseconds(Timeout.Infinite);
+                        using (var response = await noTimeoutClient.SendAsync(request, HttpCompletionOption.ResponseHeadersRead))
+                        using (var reader = new StreamReader(await response.Content.ReadAsStreamAsync()))
+                        {
+                            while (!reader.EndOfStream && !_disposed)
+                            {
+                                reader.ReadLine(); // Read the line-json update
+                                _watchCallback?.Invoke();
+                            }
```

Comment:
This might need something like an RxSubscription with a debounce? I don't know if the SecretsManager handles time debouncing for secrets changes @fabiocav ?